src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,307 @@
+import arrayFindIndex from "../../../utils/array_find_index";
+import {
+  areRangesOverlapping,
+  convertToRanges,
+} from "../../../utils/ranges";
+import request from "../../../utils/request";
+import log from "./log";
+import prepareSourceBuffer from "./media_source";
+
+interface IThumbnailInfo {
+  start: number;
+  duration: number;
+  mediaURL: string;
+}
+
+interface IThumbnailTrack {
+  thumbnailInfos: IThumbnailInfo[];
+  initURL: string;
+  codec: string;
+}
+
+/**
+ * Load needed segment data.
+ * @param {Object} thumbnails
+ * @param {HTMLMediaElement} mediaElement
+ */
+function getSegmentsData(
+  thumbnails: IThumbnailInfo[],
+  mediaElement: HTMLMediaElement
+): Observable<ArrayBuffer> {
+
+  const thumbnailsToLoad = thumbnails.filter((t) => {
+    const tRange = { start: t.start, end: t.start + t.duration };
+    const mediaRanges = convertToRanges(mediaElement.buffered);
+    return arrayFindIndex(mediaRanges, (mr) => {
+      return tRange.start >= mr.start && tRange.end <= mr.end;
+    }) === -1;
+  });
+
+  if (!thumbnailsToLoad.length) {
+    return EMPTY;
+  }
+
+  const loadedData$ = thumbnailsToLoad.map(({ mediaURL }) => {
+    return request({
+      url: mediaURL,
+      responseType: "arraybuffer",
+    }).pipe(take(1));
+  });
+
+  return observableMerge(...loadedData$).pipe(
+    map(({ value: { responseData }}) => {
+      return responseData;
+    })
+  );
+}
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * From a given video adaptation, the tools will extract a "thumbnail track",
+ * either from a trickMode track (whose light chunks are adapted from such use case)
+ * or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement: HTMLVideoElement;
+
+  private _thumbnailTrack: IThumbnailTrack;
+
+  private _setTime$: Subject<{ time: number;
+                               resolve: () => void;
+                               reject: (err: Error) => void; }>;
+  private _setTimeSubscription$: Subscription;
+
+  constructor(
+    videoElement: HTMLVideoElement,
+    trickModeTrack: Representation
+  ) {
+    // readonly
+    this._thumbnailVideoElement = videoElement;
+
+    // nullable
+    this._thumbnailTrack = this.updateThumbnailTrack(trickModeTrack);
+    this._setTime$ = new Subject();
+
+    const videoSourceInfos$ = prepareSourceBuffer(
+      this._thumbnailVideoElement,
+      this._thumbnailTrack.codec
+    ).pipe(
+      mergeMap((videoSourceBuffer) => {
+        const { initURL: init, codec } = this._thumbnailTrack;
+        return request({ url: init,
+                         responseType: "arraybuffer",
+        }).pipe(
+          mergeMap((e) => {
+            const { value: { responseData }} = e;
+            return videoSourceBuffer.appendSegment({ initSegment : responseData,
+                                                     chunk: null,
```

Comment:
I don't get the point of having a separated `initSegment` and `chunk` argument here.